File: backend/internal/models/agent.go
package models

import "encoding/json"

// Agent is a named AI persona with a bound LLM, system prompt, and tool set.
type Agent struct {
	ID                 string   `json:"id"`
	Name               string   `json:"name"`
	Description        string   `json:"description,omitempty"`
	ModelID            string   `json:"model_id"`
	SystemPrompt       string   `json:"system_prompt,omitempty"`
	Temperature        float64  `json:"temperature"`
	MaxTokens          int      `json:"max_tokens"`
	MaxIterations      int      `json:"max_iterations"`
	Tools              []string `json:"tools"`
	ToolPermissionMode string   `json:"tool_permission_mode"` // "allow_all" | "deny_all" | "manual"
	RequireApprovalFor []string `json:"require_approval_for"`
	IsEnabled          bool     `json:"is_enabled"`
	CreatedBy          string   `json:"created_by,omitempty"`
	CreatedAt          string   `json:"created_at"`
	UpdatedAt          string   `json:"updated_at"`
}

// AgentSession is a single conversation run for an agent against an optional target server.
type AgentSession struct {
	ID             string `json:"id"`
	AgentID        string `json:"agent_id"`
	UserID         string `json:"user_id"`
	ServerID       string `json:"server_id,omitempty"`
	Title          string `json:"title"`
	Status         string `json:"status"` // active | completed | failed | paused
	ContextSummary string `json:"context_summary,omitempty"`
	TokenInput     int    `json:"token_input"`
	TokenOutput    int    `json:"token_output"`
	StartedAt      string `json:"started_at"`
	EndedAt        string `json:"ended_at,omitempty"`
	CreatedAt      string `json:"created_at"`
}

// AgentMessage records one turn (user text, assistant text, or a tool call/result) within a session.
type AgentMessage struct {
	ID         string          `json:"id"`
	SessionID  string          `json:"session_id"`
	Role       string          `json:"role"` // user | assistant | tool
	Content    string          `json:"content,omitempty"`
	ToolCallID string          `json:"tool_call_id,omitempty"`
	ToolName   string          `json:"tool_name,omitempty"`
	ToolArgs   json.RawMessage `json:"tool_args,omitempty"`
	ToolResult json.RawMessage `json:"tool_result,omitempty"`
	Metadata   json.RawMessage `json:"metadata,omitempty"`
	CreatedAt  string          `json:"created_at"`
}

// ToolCall describes a single tool invocation requested by the agent LLM.
type ToolCall struct {
	Name string          `json:"name"`
	Args json.RawMessage `json:"args,omitempty"`
}

// ApprovalStatus enumerates the states of an approval request.
type ApprovalStatus string

const (
	ApprovalPending  ApprovalStatus = "pending"
	ApprovalApproved ApprovalStatus = "approved"
	ApprovalRejected ApprovalStatus = "rejected"
	ApprovalExpired  ApprovalStatus = "expired"
)

// ApprovalRequest captures a tool invocation that needs human approval.
type ApprovalRequest struct {
	ID           string          `json:"id"`
	RequestType  string          `json:"request_type"`
	Status       ApprovalStatus  `json:"status"`
	RequestedBy  string          `json:"requested_by,omitempty"`
	ToolName     string          `json:"tool_name"`
	ToolArgs     json.RawMessage `json:"tool_args,omitempty"`
	RiskLevel    string          `json:"risk_level"`
	DryRunResult json.RawMessage `json:"dry_run_result,omitempty"`
	ApprovedBy   string          `json:"approved_by,omitempty"`
	CreatedAt    string          `json:"created_at"`
}

File: frontend/src/services/agent.ts
import { apiGet, apiPost, apiPut, apiDelete } from './api';

export interface Agent {
    id: string;
    name: string;
    description?: string;
    model_id: string;
    system_prompt?: string;
    temperature: number;
    max_tokens: number;
    max_iterations: number;
    tools: string[];
    tool_permission_mode: string;
    require_approval_for: string[];
    is_enabled: boolean;
    created_by?: string;
    created_at: string;
    updated_at: string;
}

export interface AgentSession {
    id: string;
    agent_id: string;
    user_id: string;
    server_id?: string;
    title: string;
    status: string;
    context_summary?: string;
    token_input: number;
    token_output: number;
    started_at: string;
    ended_at?: string;
    created_at: string;
}

export interface AgentMessage {
    id: string;
    session_id: string;
    role: string;
    content?: string;
    tool_call_id?: string;
    tool_name?: string;
    tool_args?: any;
    tool_result?: any;
    metadata?: any;
    created_at: string;
}

export interface ApprovalRequest {
    id: string;
    request_type: string;
    status: string;
    requested_by?: string;
    tool_name: string;
    tool_args?: any;
    risk_level: string;
    dry_run_result?: any;
    approved_by?: string;
    created_at: string;
}

const api = {
    listAgents: (enabled?: boolean) =>
        apiGet<{ code: number; data: { items: Agent[]; total: number } }>(
            `/agents${enabled != null ? `?enabled=${enabled}` : ''}`
        ).then(r => r.data),
    getAgent: (id: string) =>
        apiGet<{ code: number; data: Agent }>(`/agents/${id}`).then(r => r.data),
    createAgent: (a: Partial<Agent>) =>
        apiPost<{ code: number; data: Agent }>('/agents', a).then(r => r.data),
    updateAgent: (id: string, a: Partial<Agent>) =>
        apiPut<{ code: number; data: Agent }>(`/agents/${id}`, a).then(r => r.data),
    deleteAgent: (id: string) =>
        apiDelete<{ code: number; data: any }>(`/agents/${id}`),

    createSession: (body: { agent_id: string; query: string; server_id?: string }) =>
        apiPost<{ code: number; data: AgentSession }>(`/agents/${body.agent_id}/sessions`, body).then(r => r.data),
    listSessions: (agentId?: string) =>
        apiGet<{ code: number; data: { items: AgentSession[]; total: number } }>(
            `/agents/sessions${agentId ? `?agent_id=${agentId}` : ''}`
        ).then(r => r.data),
    getSession: (id: string) =>
        apiGet<{ code: number; data: { session: AgentSession; messages: AgentMessage[] } }>(
            `/agents/sessions/${id}`
        ).then(r => r.data),
    sendMessage: (sessionId: string, message: string) =>
        apiPost<{ code: number; data: any }>(`/agents/sessions/${sessionId}/messages`, { message }),

    listApprovals: (status?: string) =>
        apiGet<{ code: number; data: { items: ApprovalRequest[]; total: number } }>(
            `/approvals${status ? `?status=${status}` : ''}`
        ).then(r => r.data),
    getApproval: (id: string) =>
        apiGet<{ code: number; data: ApprovalRequest }>(`/approvals/${id}`).then(r => r.data),
    approve: (id: string) =>
        apiPost(`/approvals/${id}/approve`, {}).then(r => r),
    reject: (id: string) =>
        apiPost(`/approvals/${id}/reject`, {}).then(r => r),
};

export { api as agentApi };
